### PR TITLE
ENH: override dweibull distribution survival and inverse survival function

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
Part of #17832 

#### What does this implement/fix?
Overrides `sf` and `isf` for the double Weibull distribution.

#### Additional information
For both functions, the available range and precision improved for positive values.

*Survival function*

image

<details>

```
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt

x=np.logspace(0, 30, 10000)
c = 1e-1
plt.loglog(x, stats.dweibull.sf(x, c), label="main")
plt.loglog(x, np.where(x > 0, 0.5 * np.exp(-abs(x)**c), 1 - 0.5 * np.exp(-abs(x)**c)), label="PR")
plt.legend()
plt.title(f"Dweibull survival function: $c={c}$")
plt.show()
```
</details>

image

<details>

```
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
from mpmath import mp

mp.dps = 100

x=np.logspace(0, 30, 10000)
@np.vectorize
def sf_mpmath(x, c):
    x = mp.mpf(x)
    c = mp.mpf(c)
    if x > 0:
        return mp.mpf(0.5) * mp.exp(-x**c)
    else:
        return mp.one - mp.mpf(0.5) * mp.exp(-(-x)**c)

def sf_pr(x, c):
    return np.where(x > 0,
                    0.5 * np.exp(-abs(x)**c),
                    1 - 0.5 * np.exp(-abs(x)**c))

c = 0.1

x = np.logspace(0, 30, 10000)
main = stats.dweibull.sf(x, c)
pr = sf_pr(x, c)
ref = np.array(sf_mpmath(x, c), np.float64)

plt.loglog(x, np.abs(main - ref)/ref, label="main", alpha=0.5)
plt.loglog(x, np.abs(pr - ref)/ref, label="PR", alpha=0.5)
plt.title(f"Dweibull survival function relative error: $c={c}$")
plt.show()
```

</details>

*Inverse survival function*

image

<details>

```
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt

def dweibull_isf(q, c):
    fac = 2. * np.where(q <= 0.5, q, 1. - q)
    fac = np.power(-np.log(fac), 1.0 / c)
    return np.where(q > 0.5, -fac, fac)

q = np.logspace(-20, -10, 1000)
plt.loglog(q, stats.dweibull.isf(q, c), label="main", ls="dashed")
plt.loglog(q, dweibull_isf(q, c), label="PR", ls="dashdot")
plt.legend()
plt.title(f"Dweibull inverse survival function: $c={c}$")
plt.show()
```
</details>

image

<details>

```
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
from mpmath import mp

mp.dps = 50

@np.vectorize
def dweibull_isf_mpmath(q, c):
    q = mp.mpf(q)
    c = mp.mpf(c)
    if q <= mp.mpf(0.5):
        fac = mp.mpf(2.) * q
    else:
        fac = mp.mpf(2.) * (mp.one - q)
    fac = (-mp.log(fac))**(mp.one / c)
    #print(fac)
    if q > mp.mpf(0.5):
        return float(-fac)
    else:
        return float(fac)

q = np.logspace(-20, 0, 1000)
c = 0.1
ref_vals = dweibull_isf_mpmath(q, c)
ref = np.array(ref_vals, np.float64)
main = stats.dweibull.isf(q, c)
pr = dweibull_isf(q, c)

plt.loglog(q, np.abs(main - ref)/main, label="main", alpha=0.5)
plt.loglog(q, np.abs(pr - ref)/main, label="PR", alpha=0.5)
plt.title(f"Dweibull inverse survival function relative error: $c={c}$")
plt.show()
```
</details>

### CODE REVIEW

**Comment by mdhaber** [comment]
The `dweibull` survival function is half the `weibull` survival function. To simplify, can you use the private functions of `weibull` as we did for `_entropy` to make this improvement? (And if the accuracy is not as good, improve the more common `weibull_min` functions instead so that both benefit.)

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -1732,6 +1732,15 @@ def _ppf(self, q, c):
...
 
+    def _sf(self, x, c):
+        Cx1 = 0.5 * np.exp(-abs(x)**c)
+        return np.where(x > 0, Cx1, 1 - Cx1)
+
+    def _isf(self, q, c):
+        fac = 2. * np.where(q <= 0.5, q, 1. - q)
+        fac = np.power(-np.log(fac), 1.0 / c)
+        return np.where(q > 0.5, -fac, fac)
+
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:1743]:
```suggestion
    def _sf(self, x, c):
        p = weibull_min._sf(x, c) / 2
        return np.where(x > 0, p, 1 - p)

    def _isf(self, q, c):
        i = (q <= 0.5)
        fac = 2. * np.where(i, q, 1 - q)
        fac = weibull_min._isf(fac, c)
        return np.where(i, fac, -fac)
```

This fails tests until `_sf` is overridden for `weibull_min`. Following the style of its `ppf`,
```python3
    def _isf(self, q, c):
        return pow(-np.log(q), 1.0/c)
```
or 
```python3
    def _isf(self, q, c):
        return (-np.log(q))**(1.0/c)
```
would be fine, too.

File: scipy/stats/tests/test_distributions.py [hunk 2]
```diff
@@ -6351,6 +6351,25 @@ def test_entropy(self):
...
+    # else:
+    #          return mp.one - mp.mpf(0.5) * mp.exp(-(-x)**c)
+    # for the inverse survival function tests, swap ref and x
+
+    @pytest.mark.parametrize('x, c, ref',
+                             [(1e20, 0.1, 1.8600379880103705e-44),
+                              (1e5, 0.5, 2.306726997904701e-138)])
+    def test_sf_isf(self, x, c, ref):
+        assert_allclose(stats.dweibull.sf(x, c), ref, rtol=5e-14)
+        assert_allclose(stats.dweibull.isf(ref, c), x, rtol=5e-14)
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:6371]: Let's convert this to a test for `weibull_min.sf`/`isf`. The test for `dweibull.sf`/`isf` can follow the style of `test_entropy`. There the comment doesn't need to be repeated, though.

**Review by mdhaber — CHANGES_REQUESTED** [review]

File: scipy/stats/tests/test_distributions.py [hunk 3]
```diff
@@ -6349,6 +6349,20 @@ def test_fit_min(self):
...
+    #     x = mp.mpf(x)
+    #     c = mp.mpf(c)
+    #     return float(mp.exp(-x**c))
+
+    @pytest.mark.parametrize('x, c, ref', [(50, 1, 1.9287498479639178e-22),
+                                           (1000, 0.8,
+                                            8.131269637872743e-110)])
+    def test_sf_isf(self, x, c, ref):
+        assert_allclose(stats.weibull_min.sf(x, c), ref, rtol=1e-14)
+        assert_allclose(stats.weibull_min.isf(ref, c), x, rtol=1e-14)
```
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:6364]: The tests for the inverse survival function fail in main.

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:6364]: Yup. You'll need to override `weibull_min._sf` with the same idea as you originally had for `dweibull._sf`.

Comment (dschmitz89) [reply scipy/stats/tests/test_distributions.py:6364]: Not sure if I understand exactly: weibull's `sf` method is already overwritten and very accurate.

File: scipy/stats/tests/test_distributions.py [hunk 4]
```diff
@@ -6349,6 +6349,20 @@ def test_fit_min(self):
         ref = np.mean(rvs), stats.skew(rvs)
         assert_allclose(res, ref)
 
+    # reference values were computed via mpmath
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:6352]: Confirmed with 
```python3
class WeibullMin(ReferenceDistribution):

    def __init__(self, *, c):
        super().__init__(c=c)

    def _pdf(self, x, c):
        return c*x**(c-mp.one)*mp.exp(-x**c)
```

**Review by mdhaber — APPROVED** [review]

File: scipy/stats/_continuous_distns.py (hunk 1 above)
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py]: Let's not accept and silently revert changes in the future, please.

**Pull request merged**